### ISSUE TITLE
build_data_loaders の augmentation 設定方法を改善

### ISSUE THREAD

**Issue by kurorosu** [kurorosu/pochidetection#401]
## Summary

`build_data_loaders()` で `hasattr(train_dataset, "_augmentation")` によるダック型チェックとプライベート属性の直接上書きを行っており, 型安全性に欠ける.

## Background

現在の実装ではデータセット構築後に `train_dataset._augmentation = augmentation` で外部からプライベート属性を設定している. `BaseCocoDataset` のコンストラクタは既に `augmentation` パラメータを受け取れるため, ファクトリ経由で渡す方が設計として適切.

## Current Implementation

```python
# training.py L173-174
if augmentation is not None and hasattr(train_dataset, "_augmentation"):
    train_dataset._augmentation = augmentation
```

- `hasattr()` による型チェックなしのダック型チェック
- プライベート属性 `_augmentation` の外部からの直接上書き
- `IDetectionDataset` インターフェースに `_augmentation` は未定義

## Proposed Changes

- `hasattr` + `_augmentation` 直接セットを `isinstance(train_dataset, BaseCocoDataset)` チェックに変更
- または `DatasetFactory` の型定義を拡張して augmentation を渡せるようにする

## Acceptance Criteria

- プライベート属性の外部からの直接上書きが解消される
- augmentation は学習データのみに適用される (既存動作維持)
- 既存テストが全て通る

## Labels

`refactor`

## Branch

`refactor/augmentation-dataset-factory`
